makeDTFromGeno <- function(geno){
  # Fix samplenames, find CHROM and POS from rownames.
  sampleNames <-  gsub("\\..*", "", basename(colnames(geno)))
  rn <- rownames(geno)
  chr <- gsub(":.*", "", rn)
  pos <- gsub("_.*", "", gsub("^.*:", "", rn))
  if(is.list(geno[1,1])){
    ul <- unlist(geno)
    vals <- paste(ul[seq(1, length(ul), by = 2)], ul[seq(2, length(ul), by = 2)], sep = ",")
    geno <- data.table(matrix(vals, ncol = ncol(geno)))
  }
  dt <- data.table(geno)
  setnames(dt, sampleNames)
  dt[, POS := as.numeric(pos)]
  dt[, CHROM := as.character(chr)]
  return(dt)
}

#' Read a vcf file to extract information relevant to XCI
#'
#' Read a given vcf file and extract minimum information required for the
#' estimation of the expression of inactivated X chromosome.
#'
#' @param vcf_file A \code{character}. The path to a vcf file.
#' @return A \code{data.table} object in long format of length sample*site.
#' @importFrom VariantAnnotation ScanVcfParam readVcf geno alt ref
#' @importFrom IRanges CharacterList
#' @importFrom Biostrings unstrsplit
#' @export
readXVcf <- function(vcf_file){
  vcf_param <- ScanVcfParam(fixed = c("ALT"), info = NA, geno = c("GT", "AD"))
  vcf <- readVcf(vcf_file, genome = "chrX", param = vcf_param)
  x = CharacterList(alt(vcf))
  x = unstrsplit(x)
  keep = which(nchar(x)==1) # Removes empty (monomorphic) and multiple (indels).
  nomono <- vcf[keep]

  # Extract allelic depth
  AD <- geno(nomono)$AD #We use unfiltered allele depth b/c filters are usually made for DNA data
  dt <- makeDTFromGeno(AD)
  dt[, ID := keep] #To filter the vcf when dt is filtered

  # Add reference and alternate alleles
  ref <- as.character(ref(nomono))
  alt <- unstrsplit(CharacterList(alt(nomono)))
  dt[, REF := ref]
  dt[, ALT := alt]
  dt <- dt[nchar(REF) == 1] #indels are a lot harder to analyse and will move the reading frame. Remove them.

  # Extract GT
  nomono <- vcf[dt$ID] # At this point it's vcf - monomorphic sites - indels on ALT - indels on REF
  GT <- geno(nomono)$GT
  dtGT <- makeDTFromGeno(GT)

  ## Trying to melt stuff
  mdt <- melt(dt, measure.vars = grep("^[A-Z]{2}[0-9]+", names(dt)), variable.name = "subject", value.name = "AD")
  mdtGT <- melt(dtGT, measure.vars = grep("^[A-Z]{2}[0-9]+", names(dtGT)), variable.name = "subject", value.name = "GT")
  dt2 <- merge(mdt, mdtGT, by = c("subject", "CHROM", "POS"))
  dt2[GT %in% c("0/0", "1/1"), AD := "NA,NA"]
  dt2[, `:=`(c("AD_ref", "AD_alt"), tstrsplit(AD, split = ","))]
  dt2[GT == "1/0",`:=`(c("AD_ref", "AD_alt"), list(AD_alt, AD_ref))]
  dt2[, AD_ref := as.numeric(AD_ref)] #will coerce "NA" into NA with a warning
  dt2[, AD_alt := as.numeric(AD_alt)]
  return(dt2)
}

#IN: the return of readXVcf, anno_file generated by dajiang's scripts. test_siteRef.tsv.anno
#OUT: dt merged with the annotations //noMono.chr*.DPR.anno//
#' Read annotation file
#'
#' Read a given annotation file and merge it with a data.table containing the
#' relevant information to exstimate inactivated X chromosome expression.
#'
#' @param dt A \code{data.table} object.
#' @param anno_file A \code{character}. The name of a file containing annotations.
#'
#' @return A \code{data.table} object that contains allelic coverage, genotype
#' and annotations at the covered SNPs.
#'
#' @export
addAnno <- function(dt, anno_file){
  anno <- fread(anno_file)
  setnames(anno, gsub("#", "", names(anno)))
  anno[, GENE := gsub("^.*:", "", ANNO)]
  anno[, ANNO := gsub(":.*$", "", ANNO)]
  anno[, CHROM := as.character(CHROM)]
  anno[, POS := as.numeric(POS)]
  anno[, ANNO_FULL := NULL]
  DPR.anno <- merge(unique(anno), dt,  by = c("CHROM", "POS", "REF", "ALT"))
  return(DPR.anno)
}
